R/elasticsearch_parsers.R

```diff
@@ -402,46 +403,58 @@ unpack_nested_data <- function(chomped_df, col_to_unpack) {
         stop(msg)
     }
     
-    # Avoid side effects
     outDT <- data.table::copy(chomped_df)
-    
-    # Get the column to unpack
     listDT <- outDT[[col_to_unpack]]
     
-    # Make each row a data.table
-    listDT <- lapply(listDT, data.table::as.data.table)
-    
-    # Remove the empty ones... important, due to data.table 1.10.4 bug
-    oldIDs <- which(sapply(listDT, nrow) != 0)
-    listDT <- listDT[oldIDs]
-    
-    # Bind them together with an ID to match to the other data
-    newDT <- data.table::rbindlist(listDT, fill = TRUE, idcol = TRUE)
-    
-    # If we tried to unpack an empty column, fail
-    if (nrow(newDT) == 0) {
+    # Check for empty column
+    if (all(lengths(listDT) == 0)) {
         msg <- "The column given to unpack_nested_data had no data in it."
         futile.logger::flog.fatal(msg)
         stop(msg)
     }
     
-    # Fix the ID because we may have removed some empty elements due to that bug
-    newDT[, .id := oldIDs[.id]]
+    listDT[lengths(listDT) == 0] <- NA
     
-    # Merge
-    outDT[, .id := .I]
-    outDT <- newDT[outDT, on = ".id"]
+    is_df <- purrr::map_lgl(listDT, is.data.frame)
+    is_na <- is.na(listDT)
+    is_atomic <- purrr::map_lgl(listDT, purrr::is_atomic)
     
-    # Remove the id column and the original column
-    outDT <- outDT[, !c(".id", col_to_unpack), with = FALSE]
+    # Bind packed column into one data.table
+    if (all(is_df | is_na)) {
+        # If the packed column contained data.frames, we need to create a row of 
+        # NAs data.table for each empty row in the packed column
+        first_nonmissing <- min(which(is_df))
+        DTNames <- names(listDT[[first_nonmissing]])
+        
+        newDT <- purrr::map(
+            listDT, 
+            function(x) {
+                if(!is.data.frame(x)) {
+                    x <- as.list(setNames(rep(NA, length(DTNames)), DTNames))
+                    x <- data.table::as.data.table(x)
+                }
+                x
+            }
+        )
+        newDT <- data.table::rbindlist(newDT, fill = TRUE)
+    } else if (all(is_atomic)) {
+        newDT <- data.table::as.data.table(unlist(listDT))
+    } else {
+        msg <- "For unpack_nested_data, col_to_unpack must be all atomic vectors or all data frames"
+        futile.logger::flog.fatal(msg)
+        stop(msg)
+    }
     
-    # Rename unpacked column if it didn't get a name
+    # Create the unpacked data.table by replicating the originally unpacked
+    # columns by the number of rows in each entry in the original unpacked column
+    group_vars <- setdiff(names(chomped_df), c(names(newDT), col_to_unpack))
+    n <- pmax(purrr::map_int(listDT, NROW), 1)
+    rest <- chomped_df[rep(1:nrow(chomped_df), n), ..group_vars, drop = FALSE]
```

Comment:
it's not obvious to me, staring at this line, what it actually does. Can you please add a comment and choose a more informative name than `rest`?

Reply:
what does `..groupvars`  do?

Reply:
also what does `drop = FALSE` do?